builder/azure/arm/template_factory.go

```diff
@@ -208,7 +232,14 @@ func GetVirtualMachineTemplateBuilder(config *Config) (*template.TemplateBuilder
 	}
 
 	if config.DiskEncryptionSetId != "" {
-		err = builder.SetDiskEncryptionSetID(config.DiskEncryptionSetId)
+		err = builder.SetDiskEncryptionSetID(config.DiskEncryptionSetId, chosenSecurityType)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if config.DiskEncryptionSetId == "" && chosenSecurityType != nil {
+		err = builder.SetDiskEncryptionWithPaaSKey(chosenSecurityType)
 		if err != nil {
 			return nil, err
 		}
```

Comment:
I think this would be a bit cleaner if we only called this function on confidential VM builds, rather than checking this conditional inside of the template builder.

```suggestion
	if config.DiskEncryptionSetId == "" && chosenSecurityType == hashiVMSDK.SecurityTypesConfidentialVM {
		err = builder.SetDiskEncryptionWithPaaSKey(chosenSecurityType)
		if err != nil {
			return nil, err
		}
```

Reply:
Ditto 👍

Reply:
@JenGoldstrich After digging back into the code I remembered why I was checking the chosenSecurityType in the builder rather than the factory. `hashiVMSDK.SecurityTypesConfidentialVM` is a const making an inline pointer deference impossible. In my latest commit, I removed this clause from the conditional and instead chose to ensure the pointer was not nil in the builder.

Will this work? I recognize you asked me to move logic out of the template builder and I did the exact opposite. I can adjust this again if you think it's still necessary.

Reply:
Yeah totally, this is still cleaner since we're handling those checks in one place, no worries and thanks for explaining!